python/tank/log.py

```diff
@@ -235,6 +237,125 @@ class LogManager(object):
+
+            # We need to close the file before renaming it (windows!)
+            if self.stream:
+                self.stream.close()
+                self.stream = None
+
+            # Before doing the rollover, check if the first file will fail at all.
+            # If it does, then it is a good thing that we checked otherwise the last
+            # backup would have been blown away before encountering the error.
+
+            # Take the scenario where there's only one backup. This means that
+            # doRollover would first delete the backup (.1) file so it can make
+            # room for the main file to be renamed to .1. However, if the main file
+            # can't be renamed, we've effectively lost 50% of the logs we had, which
+            # is not cool. Since most of the time only the first file will be locked,
+            # we will try to rename it first. If that fails right away as expected,
+            # we don't try any rollover and append to the current log file.
+            # and raise the _disable_rollover flag.
+
+            try:
+                os.rename(self.baseFilename, temp_backup_name)
+            except:
+                # It failed, so we'll simply append from now on.
+                log.debug("Can't rename main file:", exc_info=True)
+                self._handle_rename_failure("a", disable_rollover=True)
+                return
+
+            # Everything went well, so now simply move the log file back into place
+            # so doRollover can do its work.
+            try:
+                os.rename(temp_backup_name, self.baseFilename)
+            except:
+                # For some reason we couldn't move the backup in its place.
+                log.debug(
+                    "Unexpected issue renaming '%s' to '%s':" % (temp_backup_name, self.baseFilename),
+                    exc_info=True
+                )
+                # The main log file doesn't exist anymore, so create a new file.
+                # Don't disable the rollover, this has nothing to do with rollover
+                # failing.
+                self._handle_rename_failure("w")
+                return
+
+            # Now, that we are back in the original state we were in,
+            # were pretty confident that the rollover will work. However, due to
+            # any number of reasons it could still fail. If it does, simply
+            # disable rollover and append to the current log.
+            try:
+                super(LogManager._SafeRotatingFileHandler, self).doRollover()
+            except:
+                # Something probably failed trying to rollover the backups,
+                # since the code above proved that in theory the main log file
+                # should be renamable. In any case, we didn't succeed in renaming,
+                # so disable rollover and reopen the main log file in append mode.
+                log.debug("Rollover failed:", exc_info=True)
```

Comment:
again for a user, this is more helpful if it is more explicit: "Failed to roll over log file xxx"